File: Python/FCM_Function.py
import numpy as np

def fcm_function(X, C, V, m=2, eps=1e-5, maxTest=100):
    X = X.reshape(-1, 1)
    N, r = X.shape
    C= int(C)
    V = np.array(V).reshape(-1, 1)
    # print(V.shape,N,C)
    U = np.zeros((C, N))
    count = 0
    
    while True:
        for i in range(N):
            for j in range(C):
                Nominator = np.linalg.norm(X[i, :] - V[j, :])
                
                if Nominator == 0:
                    U[j, i] = 1
                    continue
                Sum = np.sum([
                    (Nominator / np.linalg.norm(X[i, :] - V[l, :])) ** (2 / (m - 1))
                    if np.linalg.norm(X[i, :] - V[l, :]) != 0 else 0
                    for l in range(C)
                ])

                U[j, i] = 1 / Sum if Sum != 0 else 1 
        
        # Normalize U
        max_U = np.max(U, axis=0)
        for i in range(N):
            if max_U[i] == 1:
                U[:, i] = (U[:, i] == max_U[i]).astype(int)
        
        # Cập nhật các trung tâm V
        W = np.zeros_like(V)
        for j in range(C):
            for i in range(r):
                W_nominator = np.sum((U[j,:] ** m) * X[:, i])
                W_denominator = np.sum(U[j,:] ** m)
                if W_denominator != 0:
                    W[j,i] = W_nominator / W_denominator
                else:
                    W[j,i] = 0
        
        # Kiểm tra độ thay đổi
        diff = np.linalg.norm(W - V)
        
        if diff <= eps or count >= maxTest:
            break
        V = W.copy()
        count += 1
    
    # Tính toán giá trị hàm mục tiêu J
    J = 0
    for i in range(N):
        for j in range(C):
            sum2 = np.linalg.norm(X[i, :] - V[j, :]) ** 2
            J += (U[j, i] ** m) * sum2

    # Xác định cụm của từng mẫu
    cum = np.argmax(U, axis=0)
    
    return V, U

File: FCM_Function.m
% FCM_Fuction returns center V, Fuzzy Partition Matrix U,
% and Objective Funtion J

% Parameter:
% 1. X: data
% 2. C: number of cluster
% 3. V: initial centers
% 4. m, eps, maxTest: optional parameters

% Usage:
% [V,U,J] = FCM_Function(X,C,V,m,eps,maxTest)

function [V,U,cum]=FCM_Function(X,C,V,m,eps,maxTest)
    [N,r]=size(X);
    count = 0;
    while (1>0)
        for i = 1:N
            for j = 1:C
                Nominator = 0;
                for k = 1:r
                    Nominator = Nominator + power(X(i,k)-V(j,k),2);
                end    
                Nominator = sqrt(Nominator);
                Sum = 0;
                for l =1:C
                    Denominator = 0;
                    for k = 1:r
                        Denominator = Denominator + power(X(i,k)-V(l,k),2);
                    end
                    Denominator = sqrt(Denominator);
                    if Denominator~=0
                        Sum = Sum + power(Nominator/Denominator, 2/(m-1));
                    else
                        Sum=Sum;
                    end
                end
                if Sum==0
                    U(j,i)=1;
                else
                U(j,i) = 1/Sum;
                end
            end
        end
        % Normalize U
        maxU=max(U);
        for i=1:N
            if maxU(i)==1
                for j=1:C
                    if U(j,i)~=maxU(i)
                        U(j,i)=0;
                    end
                end
            end
        end
        % Compute W from U
        for j = 1:C
            for i = 1:r
                W_nominator = 0;
                W_denominator = 0;
                for k = 1:N
                    W_denominator = W_denominator + power(U(j,k),m);
                    W_nominator = W_nominator + power(U(j,k),m) * X(k,i);
                end
                if (W_denominator ~= 0) 
                    W(j,i) = W_nominator / W_denominator;
                else
                    W(j,i) = 0;
                end
            end
        end
        diff = 0;
        for i = 1:C
            for j = 1:r
                diff = diff + power(W(i,j)-V(i,j),2);
            end
        end
        diff = sqrt(diff);
    
        if (diff <= eps)
            break;
        else
            for i = 1:C
                for j = 1:r
                    V(i,j)= W(i,j);
                end
            end
        end
        if(count>=maxTest)
            break;
        end
        count = count + 1;
    end
    % Compute J
    % Cluster
    maxU=max(U);
    for i=1:N
        for j=1:C
            if U(j,i)==maxU(i)
                cum(i)=j;
            end
        end
    end
     
    J = 0;
    for k=1:N
        Sum1 = 0;
        for j=1:C
            Sum2 = 0;
            for i=1:r
                Sum2 = Sum2 + power(X(k,i)-V(j,i),2);
            end
            Sum1 = Sum1 + power(Sum2,1-m);
        end
        J = J + power(Sum1,1-m);
    end
end